src/jv.h

```diff
@@ -33,7 +33,8 @@ struct jv_refcnt;
    Really. Do not play with them. */
 typedef struct {
```

Comment:
You cannot use bitfields in a public struct; their order is not defined; the compiler can order however it wants.

Reply:
If `libjq` and the program that uses `libjq` are compiled with different compilers there are no guarantees of what bits will be used, so this is likely to not work.
Even if they are compiled with the same compiler, it may still not work because the order of bitfields needs to only be consistent in individual compilation units; the compiler may feel like ordering them differently for optimisan reasons.
Bit-fields don't just subdivide the bits of a struct in the specified order, like regular fields do for bytes: they are different, and their order is not enforsed or stable. Only suitable for internal structs.

Reply:
I noticed that all compilers I tried were suspiciously generating using the same layout for bit-fields.
I just learned that apparently the SysV ABI (3.1.2; for amd64) specifies how bit-fields must be layed out in memory for public structs; they need to be layed out in order like for regular fields dividing the byte from most significant bit to least significant bit. Interesting!
Still you probably should avoid using bit fields in a public struct since standard C does not support them, and `libjq` does not only target `amd64` on UNIX-like systems (I didn't bother to check whether other standard ABIs support bit-fields in structs).

Reply:
> You cannot use bitfields in a public struct; their order is not defined; the compiler can order however it wants.

C99 says, § 6.7.2.1, constraint 10:

> 10. An implementation may allocate any addressable storage unit large enough to hold a bit-
field. If enough space remains, a bit-field that immediately follows another bit-field in a
structure shall be packed into adjacent bits of the same unit. If insufficient space remains,
whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is
implementation-defined. The order of allocation of bit-fields within a unit (high-order to
low-order or low-order to high-order) is implementation-defined. The alignment of the
addressable storage unit is unspecified.

So, yes, we can't have bit-fields in public structures when either those bits are meant to be accessed by the application or by `inline` functions or macros exported by the header.  But, as you note, _platforms_ often add their own specifications around some of these issues in their ABIs.  SysV, for example, does, and so does [Windows](https://learn.microsoft.com/en-us/cpp/c-language/c-bit-fields?view=msvc-170):

> Bit fields defined as `int` are treated as `signed`. A Microsoft extension to the ANSI C standard allows `char` and `long` types (both `signed` and `unsigned`) for bit fields. Unnamed bit fields with base type `long`, `short`, or `char` (`signed` or `unsigned`) force alignment to a boundary appropriate to the base type.

> Bit fields are allocated within an integer from least-significant to most-significant bit.

> The Windows ABI convention packs bit fields into single storage integers, and doesn't straddle storage units.

Which takes care of all concerns here for _Windows_.  I believe we'll find the same is true for Solaris/Illumos, Linux, and OS X.  I believe all of those follow SVID to some degree in their ABIs, except maybe OS X.

That said, I've no problem with using a bit of `pad_` w/o using bit fields just in case.

Reply:
Well, but what about non-amd64 architectures?

Anyway, those bit-fields should be replaced with masking and `#define JV_PAD(j) ((j.thing) & 0x7f)` macros, yes.
Also, I suggest using the most signficant bit for the boolean flag and the least significant bits for the integer instead of the other way around.

Reply:
Let's get the rest of this PR in good enough shape to rebase-and-merge and then we can resolve this comment.

Reply:
If bitfields are an issue we could use mod operations instead

Reply:
Just saw that [mongoose](https://mongoose.ws/documentation/#struct-mg_connection) uses bitfields as well. Seeing as mongoose is quite widely used and they showcase it being used on different architectures, I do not think bitfields will be an issue.